### PR TITLE
feat(python)!: allow `from_arrow` to take a generator of RecordBatches, change error type to `TypeError`

### PR DESCRIPTION
**Pull request by alexander-beedie**
Slightly extends `pl.from_arrow` constructor to allow a generator of `RecordBatch` as input (which the underlying `pa.Table.from_batches` method accepts). 

---

As well as being convenient, _might_ offer some modest memory advantages if receiving multiple batches (started looking into allowing `read_database` to take connection objects, and snowflake has a [fetch_arrow_batches](https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-api#fetch_arrow_batches) method, for example) vs requiring up-front single-batch materialisation. Need to check the pyarrow source to see how it's handled there to confirm that though.

### CODE REVIEW

File: py-polars/polars/convert.py [hunk 1]
```diff
@@ -617,12 +619,28 @@ def from_arrow(
...
+        return pl.DataFrame(
+            data=[],
+            schema=schema,
+            schema_overrides=schema_overrides,
         )
 
+    data_type = repr(type(data).__name__)
+
+    if isinstance(data, Generator):
+        if isinstance(first_batch := next(data), pa.RecordBatch):
```
Comment (stinodego) [inline py-polars/polars/convert.py:631]: This will raise a `StopIteration` if the generator is empty.

**Review by stinodego — COMMENTED** [review]
Can't we just accept an Iterable and let `pa.Table.from_batches` do the work? I'm taking a closer look here.

**Comment by stinodego** [comment]
I added a commit with what I think is a simpler implementation. What do you think? Feel free to reset / do things differently, of course, but I thought this would be the simplest way to show my thinking.

**Comment by alexander-beedie** [comment]
> Can't we just accept an Iterable and let pa.Table.from_batches do the work?

Not if we want to generate the same error when a generator containing an incorrect type is passed-in; if we don't mind being a bit inconsistent then yes.

Technically we're actually raising the wrong class of error here; should be a TypeError (as `pa.Table.from_batches` would raise), not a ValueError, but was going to fix that as a being breaking update afterwards 😅 

How about I fix it up (making it a TypeError) and simplify as Iterable at the same time, and we push it out with `0.19`? Two birds, one stone...

**Review by stinodego — APPROVED** [review]

**Comment by stinodego** [comment]
All right then, this can be merged! I'll let you do the honors :)

**Pull request merged**